pkg/storage/ingest/pusher.go

```diff
@@ -395,6 +397,10 @@ func labelAdaptersHash(b []byte, ls []mimirpb.LabelAdapter) ([]byte, uint64) {
 // PushToStorage ignores SkipLabelNameValidation because that field is only used in the distributor and not in the ingester.
 // PushToStorage aborts the request if it encounters an error.
 func (p *parallelStorageShards) PushToStorage(ctx context.Context, request *mimirpb.WriteRequest) error {
+	// We're moving Timeseries into batches, each batch has a fresh timeseries slice.
+	// We're done with the slice in the request here, the contents will live on in the batch and be freed when the batch is freed.
+	defer mimirpb.ReuseTimeseriesSliceDangerous(request.Timeseries)
```

Comment:
`Consume` is responsible for allocating from the pool, then `PushToStorage` frees back to the pool with the assumption that `Consume` is the caller. Not super familiar with the details, but can Consume do the freeing for symmetry? Like right after it calls `c.pushToStorage(...)`.

Reply:
It would definitely be cleaner to just free where it was allocated. Let's see if I can get it into that form or if I run into issues.

The trickiest part I see so far is we still have some implementation specific behavior making it hard to factor out. The parallelStoragePusher might collapse N requests into 1 batch, so you have N slices to free in that path, but only 1 slice to free in the sequential path.